### PR TITLE
Add ordering sequence to params in query

### LINKED ISSUE #72: Ordering Transaction Results Issue
https://iroha.readthedocs.io/en/main/develop/api/queries.html#result-pagination seems not to be working very well: 

by @baziorek:

"I've tried:
            pagination_meta = queries_pb2.TxPaginationMeta()
            pagination_meta.ordering.sequence = [queries_pb2.kCreatedTime, queries_pb2.kAscending]

            query = iroha.query('GetAccountAssetTransactions', account_id=account_id, asset_id=asset_name, page_size=page_size, pagination_meta=pagination_meta)
But for each try I'm seeing problem, e.g. for above:
AttributeError: Assignment not allowed to repeated field "sequence" in protocol message object."

### PR DESCRIPTION
**Pull request by turbopape**
Makes possible to add ordering sequence to Query. Fixes #72

### CODE REVIEW

File: iroha/iroha.py [hunk 1]
```diff
@@ -244,7 +244,7 @@ def command(name, **kwargs):
         return command_wrapper
 
     def query(self, name, counter=1, creator_account=None,
-              created_time=None, page_size=None, first_tx_hash=None,
+              created_time=None, page_size=None, first_tx_hash=None, ordering_sequence=None,
```
Comment (baziorek) [inline iroha/iroha.py:247]: please modify also docstring:
```
:param ordering_sequence: ...
```

File: iroha/iroha.py [hunk 2]
```diff
@@ -264,11 +264,13 @@ def query(self, name, counter=1, creator_account=None,
...
         if not creator_account:
             creator_account = self.creator_account
-        if page_size or first_tx_hash:
+        if page_size or first_tx_hash or ordering_sequence:
             pagination_meta = queries_pb2.TxPaginationMeta()
             pagination_meta.page_size = page_size
             if first_tx_hash:
                 pagination_meta.first_tx_hash = first_tx_hash
+            if ordering_sequence:
+                pagination_meta.ordering_sequence = ordering_sequence
```
Comment (baziorek) [inline iroha/iroha.py:273]: Here is problem:
```
AttributeError: 'TxPaginationMeta' object has no attribute 'ordering_sequence'
```
As I see [in documentation](https://iroha.readthedocs.io/en/main/develop/api/queries.html#result-pagination) it contains:
`ordering`, but when I try to change the name `ordering_sequence` -> `ordering`:
```
AttributeError: Assignment not allowed to field "ordering" in protocol message object.
```

File: iroha/iroha.py (hunk 1 above)
Comment (turbopape) [reply iroha/iroha.py:247]: Done :)

**Comment by baziorek** [comment]
After the changes I see that the code:
```
    ordering_sequence = [[queries_pb2.kPosition, queries_pb2.kDescending]]

    iroha2 = Iroha(user_account)
    query = iroha2.query('GetAccountAssetTransactions', account_id=user_account,
                         asset_id=asset_name, page_size=transactions_limit,
                         ordering_sequence=ordering_sequence)
```
is working and returning. But I think it may be problem with iroha c++ code - details in next post

**Comment by turbopape** [comment]
Isn't it the way to be done ? Using gRPC calls ?

**Comment by baziorek** [comment]
I've changed my entire code from: https://github.com/hyperledger/iroha-python/issues/72 by changing last two functions:
```

@trace
def get_account_transactions_different_order(user_account: str, user_private_key: str,
                                             asset_name: str, transactions_limit: int):
    ordering_sequence = [[queries_pb2.kPosition, queries_pb2.kDescending]]

    iroha2 = Iroha(user_account)
    query = iroha2.query('GetAccountAssetTransactions', account_id=user_account,
                         asset_id=asset_name, page_size=transactions_limit,
                         ordering_sequence=ordering_sequence)

    IrohaCrypto.sign_query(query, user_private_key)
    response = net.send_query(query)

    transactions = response.transactions_page_response.transactions
    for tx in transactions:
        transfer_tx = tx.payload.reduced_payload.commands[0].transfer_asset
        print(transfer_tx)


if __name__ == '__main__':
    print_paragraph('Preparing')
    create_account(user_account, user_public_key)
    create_asset(ASSET_ID)
    add_coin_to_admin(ASSET_ID, 10000)

    print_paragraph('Creating transactions')
    perform_few_transactions(user_account, 21)

    print_paragraph('Quering as usual')
    get_account_transactions(user_account, user_private_key, ASSET_ID, 21)

    print_paragraph('Quering as with different order')
    get_account_transactions_different_order(user_account, user_private_key, ASSET_ID, 1)
```
And the output for different ordering is strange:
```
---------- Quering as usual : ----------
	Entering "get_account_transactions"
src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 1"
amount: "1"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 2"
amount: "2"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 3"
amount: "3"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 4"
amount: "4"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 5"
amount: "5"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 6"
amount: "6"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 7"
amount: "7"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 8"
amount: "8"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 9"
amount: "9"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 10"
amount: "10"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 11"
amount: "11"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 12"
amount: "12"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 13"
amount: "13"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 14"
amount: "14"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 15"
amount: "15"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 16"
amount: "16"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 17"
amount: "17"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 18"
amount: "18"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 19"
amount: "19"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 20"
amount: "20"

src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 21"
amount: "21"

	Leaving "get_account_transactions"

---------- Quering as with different order : ----------
	Entering "get_account_transactions_different_order"
src_account_id: "admin@test"
dest_account_id: "a@test"
asset_id: "lemurcoin#test"
description: "tx 20"
amount: "20"

	Leaving "get_account_transactions_different_order"
```
**As You see returning last transaction (limit 1 and ordering `kDescending`) is not returning last transaction ("tx 21" should be returned, but "tx 20" is returned instead).**

**Comment by baziorek** [comment]
> Isn't it the way to be done ? Using gRPC calls ?
I think that Your work is correct. But the problem probably is in iroha-c++ code. It would be good if somebody who knows iroha and iroha-python taka a look before the issue would be closed.

**Comment by turbopape** [comment]
maybe a limit condition related to kPosition with regards to Asc / Desc?

**Comment by baziorek** [comment]
> maybe a limit condition related to kPosition with regards to Asc / Desc?

Unfortunately I don't know why it is like that. But according to documentation: https://iroha.readthedocs.io/en/main/develop/api/queries.html#result-pagination it looks for me as bug. Maybe @LiraLemur would know it it is not a bug in iroha-core code?

**Comment by turbopape** [comment]
You're right, thank you so much :)

**Comment by LiraLemur** [comment]
There seems to be some peculiarity in the main code, I think - similar difficulties occurred in JS library too but were fixed. 

@stepanLav maybe you could comment a little more on this?

**Comment by stepanLav** [comment]
Yeah, it looks like it's really a bug in Iroha.
I try to call it through [JS](https://github.com/hyperledger/iroha-javascript) library and get the same result:

```
const queryService = new QueryService(
  IROHA_ADDRESS,
  grpc.credentials.createInsecure()
)

queries.getAccountAssetTransactions({
  privateKey: adminPriv,
  creatorAccountId: adminId,
  queryService,
  timeoutLimit: 5000
}, {
  accountId: adminId,
  assetId: 'coin#test',
  pageSize: 1,
  firstTxHash: undefined,
  ordering: {
    field: 1,
    direction: 1
  }
}).then(a => {
  console.log(a);
  (a as any).transactionsList.forEach(element => {
    element.payload.reducedPayload.commandsList.forEach(element => {
      console.log(element.transferAsset)
    })
  })
}).catch(e => console.error(e))
```

output:
```
{
  transactionsList: [ { payload: [Object], signaturesList: [Array] } ],
  allTransactionsSize: 42,
  nextTxHash: '637de0e7321b40627b2152bad651f4da1a26d11d47cf1c960c2034313eaa7c69'
}
{
  srcAccountId: 'alice@test',
  destAccountId: 'a@test',
  assetId: 'coin#test',
  description: 'tx 20',
  amount: '20'
}
```

**Review by stepanLav — APPROVED** [review]

**Comment by turbopape** [comment]
Thanks guys. Does it make sense to report this issue on the main Iroha channel with reference to this discussion? What do you think ?

**Comment by baziorek** [comment]
> Thanks guys. Does it make sense to report this issue on the main Iroha channel with reference to this discussion? What do you think ?

I don't know the answer, but before merge it should be repaired:
obraz
When You click "Details" You will see commands how to correct its.

**Comment by turbopape** [comment]
Hello, that's done :)

**Review by baziorek — APPROVED** [review]

**Comment by turbopape** [comment]
Hey guys so are there any actions I should undertake for this PR tp be merged, or shall we wait for main Iroha to sort the upstream issue?
Many thanks 😊

**Comment by baziorek** [comment]
> Hey guys so are there any actions I should undertake for this PR tp be merged, or shall we wait for main Iroha to sort the upstream issue?
> Many thanks blush

I think @LiraLemur knows the answer: @LiraLemur I've checked its and also @stepanLav and everything is corrected, so maybe it can be merged?

**Review by stepanLav — APPROVED** [review]

**Comment by turbopape** [comment]
Hey guys, so, anything needed from my side for the merge ?
thanks :)

fekr.tech
Thinking Software
Rafik Naccache
Partner
Skype: rafik_naccache
Google Cloud Partner
<https://cloud.withgoogle.com/partners/detail/?id=CIGAgICAgID7LA%3D%3D&language=en>
upWork <https://www.upwork.com/fl/turbopape> - github
<https://github.com/turbopape>- linkedin
<https://www.linkedin.com/in/rafiknaccache/>
Cell: +216 51 78 77 00

"Any sufficiently advanced technology is indistinguishable from magic"
Clarke's 3rd Law

File: iroha/iroha.py [hunk 3]
```diff
@@ -264,11 +266,18 @@ def query(self, name, counter=1, creator_account=None,
             created_time = self.now()
         if not creator_account:
             creator_account = self.creator_account
-        if page_size or first_tx_hash:
+        if page_size or first_tx_hash or ordering_sequence:
```
Comment (kuvaldini) [inline iroha/iroha.py:269]: Why to combine these with OR and then add one more selective if?

**Comment by turbopape** [comment]
I've removed that silly OR :)

**Comment by LiraLemur** [comment]
@turbopape Thank you! Merged!

**Comment by turbopape** [comment]
Thank you so much :)

fekr.tech
Thinking Software
Rafik Naccache
Partner
Skype: rafik_naccache
Google Cloud Partner
<https://cloud.withgoogle.com/partners/detail/?id=CIGAgICAgID7LA%3D%3D&language=en>
upWork <https://www.upwork.com/fl/turbopape> - github
<https://github.com/turbopape>- linkedin
<https://www.linkedin.com/in/rafiknaccache/>
Cell: +216 51 78 77 00

"Any sufficiently advanced technology is indistinguishable from magic"
Clarke's 3rd Law

**Pull request merged**